lib/legacy/legacy-data-mixin.js

```diff
@@ -0,0 +1,150 @@
+*/
+
+import { Class } from './class.js';
+import { Polymer } from '../../polymer-legacy.js';
+import { dedupingMixin } from '../utils/mixin.js';
+
+const UndefinedArgumentError = class extends Error {
+  constructor(message, arg) {
+    super(message);
+    this.arg = arg;
+    this.name = this.constructor.name;
+    // Affordances for ensuring instanceof works after babel ES5 compilation
+    // TODO(kschaaf): Remove after polymer CLI updates to newer Babel that
+    // sets the constructor/prototype correctly for subclassed builtins
+    this.constructor = UndefinedArgumentError;
+    this.__proto__ = UndefinedArgumentError.prototype;
+  }
+};
+
+/**
+ * Wraps effect functions to catch `UndefinedArgumentError`s and warn.
+ * 
+ * @param {Object=} effect Effect metadata object
+ * @param {Object=} fnName Name of user function, if known
+ * @return {?Object} Effect metadata object
+ */
+function wrapEffect(effect, fnName) {
+  if (effect && effect.fn) {
+    const fn = effect.fn;
+    effect.fn = function() {
+      try {
+        fn.apply(this, arguments);
+      } catch (e) {
+        if (e instanceof UndefinedArgumentError) {
+          console.warn(`Argument '${e.arg}'${fnName ?` for method '${fnName}'` : ''} was undefined. Ensure it has an undefined check.`);
+        } else {
+          throw e;
+        }
+      }
+    };
+  }
+  return effect;
+}
+
+/**
+ * Mixin to selectively add back Polymer 1.x's `undefined` rules
+ * governing when observers & computing functions run based
+ * on all arguments being defined (reference https://www.polymer-project.org/1.0/docs/devguide/observers#multi-property-observers).
+ * 
+ * When loaded, all legacy elements (defined with `Polymer({...})`)
+ * will have the mixin applied. The mixin only restores legacy data handling
+ * if `_legacyUndefinedCheck: true` is set on the element's prototype.
+ * 
+ * This mixin is intended for use to help migration from Polymer 1.x to
+ * 2.x+ by allowing legacy code to work while identifying observers and
+ * computing functions that need undefined checks to work without
+ * the mixin in Polymer 2.
+ *
+ * @mixinFunction
+ * @polymer
+ * @summary Mixin to selectively add back Polymer 1.x's `undefined` rules
+ * governing when observers & computing functions run.
+ */
+export const LegacyDataMixin = dedupingMixin(superClass => {
+
+  /**
+   * @constructor
+   * @extends {superClass}
+   * @unrestricted
+   * @private   */
+  class LegacyDataMixin extends superClass {
+    /**
+     * Overrides `Polymer.PropertyEffects` to add `undefined` argument
+     * checking to match Polymer 1.x style rules
+     * 
+     * @param {!Array<!MethodArg>} args Array of argument metadata
+     * @param {string} path Property/path name that triggered the method effect
+     * @param {Object} props Bag of current property changes
+     * @return {Array<*>} Array of argument values
+     * @private
+     */
+    _marshalArgs(args, path, props) {
```

Comment:
Did you consider altering this method to get the info needed to present a good error? I'm concerned that since the intension is to warn, but the code throws here, we're depending on catching that where it's called.

Reply:
Ok, disregard that. It's clear that we need to abort the effect of marshaling the arguments and throwing let's us do this.